### DISCUSSION TITLE
Loading a model that contains a custom pipeline component

### DISCUSSION THREAD

**Discussion by clcarver1130** [explosion/spaCy#8088] [Help: Coding & Implementations] [answered]
Hi! Struggling to load a spacy model that contains a custom pipeline component with the new v3 format. My component simply sets a new Span extension to `True`. I build the component as shown below:

```
# Load model:
nlp = spacy.load('model_training/testing_model')
print(f'Starting pipeline objects: {nlp.pipe_names}')

# EntityRuler:
ruler = nlp.add_pipe('entity_ruler', before='ner')
ruler.add_patterns(patterns)

# Custom Pipeline Component
@Language.component("from_pattern")
def from_pattern(doc):
    Span.set_extension('from_pattern', default=False)
    for ent in doc.ents:
        ent._.from_pattern = True
    return doc
nlp.add_pipe('from_pattern', name='pattern_flag', after='entity_ruler')

# Save to disk:
print(f'Ending pipeline objects: {nlp.pipe_names}')
nlp.to_disk('model_training/testing_model')
```

I save the nlp object with the new component `to_disk` and confirm that the component is in the config file. See screenshots below:
![Screen Shot 2021-05-13 at 4 05 12 PM](https://user-images.githubusercontent.com/35936027/118187802-124ffc00-b405-11eb-8978-9dad5aee6ad9.png)
![Screen Shot 2021-05-13 at 4 05 25 PM](https://user-images.githubusercontent.com/35936027/118187808-1419bf80-b405-11eb-80ca-125a2f863a3a.png)

But when I try to load the model I get the following error:

![Screen Shot 2021-05-13 at 4 08 35 PM](https://user-images.githubusercontent.com/35936027/118188077-7a9edd80-b405-11eb-9337-c65cce7424cc.png)

Any help would be greatly appreciated!

**Comment by polm** [comment, marked as answer]
spaCy looks your component up by the name "from_pattern", so that has to be in the registry when the component is loaded. You can put that in the registry by importing the definition of the component with the decorator. 

Your sample code doesn't have an `if __name__ == "__main__"` section, so you can't import it safely without some changes, but if you copy/paste the component definition before loading your pipeline that should work.

**Reply by clcarver1130** [reply]
Thanks for the reply! If I'm understanding you correctly I should structure like below then? The issue I'm running into now is that it doesn't seem to remember the 'from_pattern' extension. Do I need to define that a second time?

Code Sample:
```
if __name__ == "__main__":

    text = 'text placeholder'

    # Custom Spacy functions:
    @Language.component("from_pattern")
    def from_pattern(doc):
        Span.set_extension('from_pattern', default=False)
        for ent in doc.ents:
            ent._.from_pattern = True
        return doc

    nlp = spacy.load('model_training/testing_model')

    doc = nlp(text)

    for ent in doc.ents:
        print(ent)
        print(ent._.pattern_flag)
```

New Error:
```
Traceback (most recent call last):
  File "model_training/test_pattern.py", line 23, in <module>
    print(ent._.pattern_flag)
  File "/Users/Cameron/miniconda3/envs/ibchallenge/lib/python3.8/site-packages/spacy/tokens/underscore.py", line 32, in __getattr__
    raise AttributeError(Errors.E046.format(name=name))
AttributeError: [E046] Can't retrieve unregistered extension attribute 'pattern_flag'. Did you forget to call the `set_extension` method?
```

**Reply by polm** [reply]
Ah, no, that's not the structure I was suggesting, and now you have another problem. 

Short answer: If this is your only script for loading, you can put this line at the top (after `text =`) and it should work.

    Span.set_extension('from_pattern', default=False)

You also need to fix the issue that you are checking for `pattern_flag` but you have created the `from_pattern` extension. These should have the same name. (You could leave the `set_extension` call in the same place, but it only needs to be called once, so it makes sense to have it outside the component. Note that [in this example](https://spacy.io/usage/processing-pipelines#component-example3) `set_extension` is called in `__init__`, which is only called once, and not `__call__`, which is called for every doc.)

Longer answer: 

In order to get your pipeline set up things need to be executed at the right time. This is a little confusing and there's interaction with several execution controlling systems in Python, so I'm going to explain them all here.

Executing the decorator adds your component to the registry. The decorator is executed when the function is defined. In your code above the code definition is inside the main block, so the function is not defined on an `import` call. 

To explain the function of the `__main__` block. The code in the block is only called if you execute that script like `python script.py`. If you call `import script` then the code in that block won't execute, but all the other code will. In your case you want to use your component without running your training code, so the training code needs to be inside the block, but you *do* want to execute the decorator, so that needs to be outside the block.

Second, decorators. A decorator is just syntactic sugar for calling a function on a function you're defining. So these are equivalent:

```
@deco
def myfunc():
    ...
```
    
```
def myfunc():
    ....

deco(myfunc)
```

In this case the `Language.component` decorator is what adds your function to the registry, so that needs to be called in order for spaCy to find your code. 

It might be easiest for you to have three files, one defining your component, one your training code, and one the loading code. Something like this:


```
# mycomponent.py

Span.set_extension(...)

@Language.factor("from_pattern")
def mycomponent():
     ...

# train.py
from mycomponent import mycomponent

# put your training loop here

# load.py
from mycomponent import mycomponent

# put your loading code here
```

**Reply by morankyle** [reply]
Near the end of your answer here you reference his training loop, so if I'm using Spacy v3 with the config file how could I fix this problem? I asked this question in more detail here but this seems very similar https://github.com/explosion/spaCy/discussions/8414

**Reply by morankyle** [reply]
Figured this out by importing the function from the python file
